### PR TITLE
pass NetworkPluginDir to kubenet

### PR DESCRIPTION
**Pull request by freehan**
to unblock GCI kubenet integration #26379

cc: @andyzheng0831

### CODE REVIEW

File: pkg/kubelet/network/kubenet/kubenet_linux.go [hunk 1]
```diff
@@ -77,28 +77,29 @@ type kubenetNetworkPlugin struct {
...
+		cniPath:     cniPath,
 	}
 }
 
 func (plugin *kubenetNetworkPlugin) Init(host network.Host, hairpinMode componentconfig.HairpinMode) error {
 	plugin.host = host
 	plugin.hairpinMode = hairpinMode
 	plugin.cniConfig = &libcni.CNIConfig{
-		Path: []string{DefaultCNIDir},
+		Path: []string{DefaultCNIDir, plugin.cniPath},
```
Comment (bprashanth) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: do we care about order if both directories have plugins?

File: pkg/kubelet/network/kubenet/kubenet_linux.go [hunk 2]
```diff
@@ -77,28 +77,29 @@ type kubenetNetworkPlugin struct {
 	hairpinMode     componentconfig.HairpinMode
 	hostPortMap     map[hostport]closeable
 	iptables        utiliptables.Interface
+	cniPath         string
```
Comment (bprashanth) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: please add a comment either here or above newplugin about what this field is and what the default is/how it interacts with defaultcniplugindir and which takes precedence.

File: pkg/kubelet/network/kubenet/kubenet_linux.go (hunk 1 above)
Comment (freehan) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: CNI search for plugins sequentially. Maybe I should put the cniPath before DefaultCNIDir?

File: pkg/kubelet/network/kubenet/kubenet_linux.go (hunk 2 above)
Comment (freehan) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: fixed

Comment (freehan) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: Or I can rename it to extraCNIPath

File: pkg/kubelet/network/kubenet/kubenet_linux.go (hunk 1 above)
Comment (bprashanth) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: What happens when we stack plugins today? is it safer to simply ditch the defaultcnidir if there's a user supplied path? 

If we're stacking them and assuming only one will be set, suggest a bold warning if both contain cni configs or something.

**Comment by bprashanth** [comment]
im fine with this. I suppose there was always the possibility of running multiple plugins, even without multiple cni paths. If we can detect this, I'd like a bold warning of some sort.

**Comment by andyzheng0831** [comment]
Could you give an instruction for how to use this change? For example, what kubelet flag should be set to the desired path?

**Comment by freehan** [comment]
kubenet has the luxury of controlling which type of plugin to run. The extraCNIPath just add an option for users to specify where they put the binaries. 

If extraCNIPath is specified, kubenet will search plugins in it first.

**Comment by freehan** [comment]
It is controlled by kubelet flag `--network-plugin-dir`.

**Comment by bprashanth** [comment]
I think you have a build failure

**Comment by bprashanth** [comment]
> kubenet has the luxury of picking which type of plugin to run. The extraCNIPath just add an option for users to specify where they put the binaries.
If extraCNIPath is specified, kubenet will search plugins in it first.

This is not really upto kubenet right? the order of that list is important (https://github.com/containernetworking/cni/blob/master/libcni/api.go#L46 is just going to take the first matching plugin in the first path and exec it, IIUC). 

I think we should be consistent with that cni plugins do. From a cursory glance at the code it looks like the others prioritize default over vendor, maybe follow the same pattern?

**Comment by bprashanth** [comment]
LGTM, please update the docs when you have time

**Pull request merged**